platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/interfaces/ResolutionFailure.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.resolution.failure.interfaces;
+
+/**
+ * Represents any failure that can occur during variant selection and
+ * contains contextual information specific to that type of failure.
+ *
+ * @implSpec This interface is meant only to be extended by other interfaces, it should not
+ * be implemented directly.  Concrete types implementing this interface should be immutable data classes
+ * without nullable fields.  They must also be serializable by the configuration cache.
```

Comment:
It is true, as these exceptions can be thrown when running the `dependencyInsight` report.  There is one test that checks for this.

I've changed those resolve state types to metadata, so everything should be fine with those failures.

Reply:
It still gives me a weird feeling that CC serializing these internal dependency resolution types -- even if they're just the immutable metadata.

I question why these failures are present in the final thrown exception at all

Reply:
In case someone wants to catch the exception and process it.  Unless there is a specific reason to avoid it, adding structured contextual information to exceptions is usually a good thing.

These failure types are going to be reported to Problems soon, as well - though this wouldn't require them being present on the exception itself.  We can discuss the need (or lack thereof) for this data later, I'm not going to remove it here.

Reply:
We do not need to fix it here, but if we actually expect people (especially via the problems API) to introspect this error, we should _not_ expose internal dependency management types.

Reply:
It would be great to have some way of marking internal dependency management types as "internal-internal" vs. just the standard "internal" packaging.  These Metadata types looked safe to me to use.

Reply:
These are "internal", not "internal internal", but your argument was for this metadata to be used in the problems API. 

We shouldn't even expose "internal" types there.